### PR TITLE
Record Serialisation for Reports

### PR DESCRIPTION
**Pull request by garethpotter**
Adds optional support for including records when serialising reports. Also fixes a bug in the JSON serialiser.

### CODE REVIEW

File: src/thingset_txt.c [hunk 1]
```diff
@@ -253,7 +253,25 @@ static int txt_serialize_value(struct thingset_context *ts,
             pos = snprintf(buf, size, "null,");
         }
         else if (object->type == THINGSET_TYPE_RECORDS) {
-            pos = snprintf(buf, size, "%d,", object->data.records->num_records);
+#ifdef CONFIG_THINGSET_REPORT_RECORD_SERIALISATION
```
Comment (martinjaeger) [inline src/thingset_txt.c:256]: Same here: `IS_ENABLED` will help with readability.

File: tests/protocol/src/txt.c [hunk 2]
```diff
@@ -584,9 +588,13 @@ ZTEST(thingset_txt, test_export_subset)
         "{"
         "\"t_s\":1000,"
         "\"Types\":{\"wBool\":true},"
+        "\"Records\":2,"
         "\"Nested\":{\"rBeginning\":1,\"Obj2\":{\"rItem2_V\":2.2}}"
         "}";
 
+    // 0xfffff7d6e208
+    // "{\"t_s\":1000,{\"wBool\":true},\"Records\":2,{\"rBeginning\":1,{\"rItem2_V\":2.2},"
+
```
Comment (martinjaeger) [inline tests/protocol/src/txt.c:597]: Leftover from testing?

File: Kconfig.thingset [hunk 3]
```diff
@@ -36,6 +36,21 @@ config THINGSET_NUM_JSON_TOKENS
...
+config THINGSET_BINARY_MAX_DEPTH
+	int "Max depth of hierarchy serialisable by the binary serialiser."
+	default 4
+	help
+	  Configure the maximum depth of a node hierarchy that can be serialised by
+	  the binary serialiser. The default of 4 is usually suitable for most use
+	  cases, but you may wish to increase this in certain scenarios.
+
+config THINGSET_REPORT_RECORD_SERIALISATION
+	bool "Enable full record serialisation in reports."
```
Comment (martinjaeger) [inline Kconfig.thingset:48]: Please use AE to be consistent with the rest of the code.

```suggestion
config THINGSET_REPORT_RECORD_SERIALIZATION
	bool "Enable full record serialization in reports."
```

**Review by martinjaeger — CHANGES_REQUESTED** [review]
Looks good, only got very minor comments.

Currently the new report serialization is not covered in the tests. Could you add a test case with that Kconfig enabled? It could even be in a separate file if it would otherwise lead to too many ifdefs in the current tests.

**Review by martinjaeger — APPROVED** [review]
Thanks for the update. Ready to be merged as soon as we have some unit tests covering the new feature.

File: src/thingset_bin.c [hunk 4]
```diff
@@ -703,6 +750,42 @@ inline void thingset_bin_setup(struct thingset_context *ts, size_t rsp_buf_offse
                            ts->rsp_size - rsp_buf_offset, 1);
 }
 
+#ifdef CONFIG_THINGSET_PROGRESSIVE_IMPORT_EXPORT
```
Comment (martinjaeger) [inline src/thingset_bin.c:753]: Doesn't this belong to the other PR?

**Review by martinjaeger — CHANGES_REQUESTED** [review]

File: src/thingset_bin.c (hunk 4 above)
Comment (garethpotter) [reply src/thingset_bin.c]: Yes, sorry, I've got 3 things on the go as you can see and I made a bit of a mess. Think it's OK now.

**Pull request merged**